pkg/provider/identity.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the Apache 2.0 license.
+*/
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"os"
+	"regexp"
+
+	azaciv2 "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerinstance/armcontainerinstance/v2"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	armcontainerservice "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice"
+	armmsi "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/msi/armmsi"
+	azidentity "github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	v1 "k8s.io/api/core/v1"
+)
+
+func SetContainerGroupIdentity(ctx context.Context, identity *armmsi.Identity, identityType azaciv2.ResourceIdentityType, containerGroup *azaciv2.ContainerGroup) {
+	if identity == nil || identityType != azaciv2.ResourceIdentityTypeUserAssigned {
+		return
+	}
+
+	cgIdentity := azaciv2.ContainerGroupIdentity{
+		Type: &identityType,
+		UserAssignedIdentities: map[string]*azaciv2.UserAssignedIdentities{
+			*identity.ID: &azaciv2.UserAssignedIdentities{
+				PrincipalID: identity.Properties.PrincipalID,
+				ClientID: identity.Properties.ClientID,
+			},
+		},
+	}
+
+	log.G(ctx).Infof("setting managed identity based imageRegistryCredentials\n")
+	containerGroup.Identity = &cgIdentity
+}
+
+func (p *ACIProvider) GetAgentPoolKubeletIdentity(ctx context.Context, pod *v1.Pod) (*armmsi.Identity, error) {
```

Comment:
It would be great if we can get the identity direct using [managedCluster](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/containerservice/armcontainerservice/models.go#L2107) SDK

Reply:
In order to get cluster-info directly using armcontainerservice (managedClusters) sdk, we need to make Get call with the clusterName. ACI Provider doesn't know the cluster name, only the cluster fqdn, so we have to filter.
The GET options are not specified right now: [ManagedClustersClientGetOptions ](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/containerservice/armcontainerservice/models.go#L2075)